core/src/ProblemManager.cc

```diff
@@ -229,27 +266,49 @@ void ProblemManager::calculateNodalMomentum(
     int space_dim = d_mesh->spatialDimension();
     int nodes_per_cell = d_mesh->nodesPerCell();
     int node_id = 0;
+    int num_nodes = d_mesh->totalNumNodes();
+
+    std::vector<std::vector<std::array<double,3> > > node_p_local(
+        d_thread_count, std::vector<std::array<double,3> > (
+	num_nodes, std::array<double,3>() ) );
 
-    // Reset the momentum
+    // Reset the momentum.
     for ( auto& mom : node_p )
         std::fill( mom.begin(), mom.end(), 0.0 );
-
-    // Update the momentum
-    for ( auto& p : d_particles )
+  
+    // Update the momentum.
+    #pragma omp parallel for num_threads(d_thread_count)
+    for ( std::size_t i = 0; i < d_particles.size(); ++i )
     {
+	auto& p = d_particles.at(i);
+	int th = omp_get_thread_num();
+
         // Calculate momentum.
         for ( int n = 0; n < nodes_per_cell; ++n )
         {
             node_id = p.node_ids[n];
 
             for ( int d = 0; d < space_dim; ++d )
             {
-                node_p[node_id][d] +=
+                node_p_local[th][node_id][d] +=
```

Comment:
Are these arrays all initialized to zero before the summation starts?

Reply:
yes, they are automatically initiallized to zero, I just double checked to make sure.

Reply:
so that is true in debug build but likely not true in a release build - you will have to manually reset to 0